[new component] auto-complete

yongxu opened the pull request:
This component is the latest update from #1985

I decided to open a new PR because of the name change.

Known bug:
from @vijayrawatsan:

> Lets say you type in 'h' - > 4 results
> After that you type 'e' -> 1 result
> 
> No you press backspace to remove letter 'e'. Now the dropdown should show 4 results but it only show 1 result, rest is empty space in dropdown.
> 
> This is fixed when I turn animated: false while creating menu.

This might be caused by bug from src/menus/menu.jsx

@oliviertassinari the api has been changed to:

```
dataSource // for list proprity

/* function(searchText,prevDataSource) -> if return a list, 
   then replace dataSource with that list, 
   otherwise use the dataSource from props */
onUpdateInput 

//and
onNewRequest(searchText,index,dataSource) //if no match then index = null
```

Please merge, it should be ready! :)

yongxu committed:
[new component] auto-complete

oliviertassinari commented on docs/src/app/components/pages/components/auto-complete.jsx:
```diff
@@ -0,0 +1,80 @@
+    return (
+      <ComponentDoc
+        name="Auto Complete"
+        componentInfo={[{
```
Could you add more information as for instance here https://github.com/callemall/material-ui/blob/master/docs/src/app/components/pages/components/dialog.jsx#L31

yongxu replied:
I'd love to, just quite busy recently. I can do it this friday. I thought maybe we should push it first and then finish it? Since the component is usable, I wish more people can try it out before the document is ready.

oliviertassinari commented on src/auto-complete.jsx:
```diff
@@ -0,0 +1,296 @@
+
+
+    let requestsListPromise = new Promise((resolve, reject) => {
+      let list = this.props.onUpdateInput(searchText, this.state.requestsList);
```
I think that expecting `onUpdateInput` to return something doesn't fit into the flux pattern. Hence I think that we should remove it.
@shaurya947 What do you think?

yongxu replied:
This isn't flux pattern, but if you just want to call an external API and not going to change anything, I think this will be a easier way to do.
Return is optional anyways, why not let themselves to decide?

shaurya947 replied:
It does feel a little weird to call a function in the owner from the ownee and have it return something to the ownee.

@yongxu I know this is taking a little long, but seriously, thanks for your patience -- we're almost there! Could you think of another way to do this? Maybe update the state inside the owner whenever `onUpdateInput` is called, and pass the updated state using some prop?

yongxu replied:
@shaurya947 I see your point. Indeed fetch and ajax use either callback or promise. This return doesn't seems that useful.

My thought was only this only happens when the ```dataSource``` doesn't exist, so it shouldn't be used for flux, and will not take the return at all.

My thought was if you want a short filter implementation with preloaded data, or work with simple request, this can be very handy to use.

I am not sure if I have another better way to do this, my thought was this can be quite useful, but not necessary. I am cool if you guys think it might be better to remove it, or I can make some change if there is a better way to do it.

oliviertassinari replied:
> My thought was if you want a shorter filter implementation with preloaded data, or work with simple request, this can be very handy to use

I understand your point. That's less boilerplate and more straightforward to use.
However, I think that having two ways to do the same thing is wrong.
First it's more code to maintain and second if you use the first more limited solution and at some point you need more power, you will have to refactor, not great.

yongxu replied:
@oliviertassinari I think you are right, maybe we should remove it.
Going to work on it, Thanks for the advice.

oliviertassinari replied:
Ok thanks! We will merge then :+1:

shaurya947 replied:
Good decision guys, looking forward to this being merged :)

yongxu committed:
revision: removed onUpdateInput callback return.
non-unidirectional data flow pattern is removed, in order to fit flux
model.
Now dataSource accept object {key:component} pattern
Improved performance.

yongxu commented:
@oliviertassinari @shaurya947 It should be ready now :)

oliviertassinari commented on docs/src/app/components/raw-code/auto-complete-code.txt:
```diff
@@ -0,0 +1,45 @@
+          <AutoComplete
```
Could you remove the extra space?

oliviertassinari commented:
@yongxu The API look good to me :+1:. Could you have a look at my comment? I will merge then

oliviertassinari commented:
@yongxu Thanks for your effort!

yongxu commented:
Thanks for all the review @oliviertassinari :+1:

vijayrawatsan commented:
@yongxu @oliviertassinari @shaurya947 :+1: 
Finally its merged.

shaurya947 commented:
Yes!
